File: src/leetcode/medium/lm_692.java
package leetcode.medium;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class lm_692 {
    private List<String> topKFrequent(String[] words, int k) {
        HashMap<String, Integer> map = new HashMap<>();
        for (String s : words)
            map.put(s, map.getOrDefault(s, 0) + 1);
        return map
                .entrySet()
                .stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed().thenComparing(Map.Entry::getKey))
                .limit(k)
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
    }
}

File: src/leetcode/medium/lm_692.kt
package leetcode.medium

fun main() {
    println(topKFrequent(arrayOf("i", "love", "leetcode", "i", "love", "coding"), 2).joinToString())
    println(topKFrequent(arrayOf("i", "love", "leetcode", "i", "love", "coding"), 2).joinToString())
    println(
        topKFrequent(
            arrayOf("the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"),
            4
        ).joinToString()
    )
}

private fun topKFrequent(words: Array<String>, k: Int): List<String> =
    HashMap<String, Int>()
        .apply { words.forEach { this[it] = getOrDefault(it, 0) + 1 } }
        .entries
        .asSequence()
        .sortedWith(compareByDescending<MutableMap.MutableEntry<String, Int>> { it.value }.thenBy { it.key })
        .take(k)
        .map { it.key }
        .toList()

private fun topKFrequent2(words: Array<String>, k: Int): List<String> {
    val hashmap = HashMap<String, Int>()
    words.forEach { hashmap[it] = hashmap.getOrDefault(it, 0) + 1 }
    return hashmap.entries.sortedWith(compareByDescending<MutableMap.MutableEntry<String, Int>> { it.value }.thenBy { it.key })
        .take(k).map { it.key }
}

